cpp/include/raft/comms/comms.hpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+
+namespace raft {
+namespace comms {
+
+typedef unsigned int request_t;
+enum datatype_t { CHAR, UINT8, INT, UINT, INT64, UINT64, FLOAT, DOUBLE };
+enum op_t { SUM, PROD, MIN, MAX };
+
+/**
+ * The resulting status of distributed stream synchronization
+ */
+enum status_t {
+  commStatusSuccess,  // Synchronization successful
+  commStatusError,    // An error occured querying sync status
+  commStatusAbort     // A failure occurred in sync, queued operations aborted
+};
+
+class comms_iface {
+ public:
+  virtual ~comms_iface();
+
+  virtual int getSize() const = 0;
+  virtual int getRank() const = 0;
+
+  virtual std::unique_ptr<comms_iface> commSplit(int color, int key) const = 0;
+
+  virtual void barrier() const = 0;
+
+  virtual status_t syncStream(cudaStream_t stream) const = 0;
+
+  virtual void isend(const void* buf, int size, int dest, int tag,
+                     request_t* request) const = 0;
+
+  virtual void irecv(void* buf, int size, int source, int tag,
+                     request_t* request) const = 0;
+
+  virtual void waitall(int count, request_t array_of_requests[]) const = 0;
+
+  virtual void allreduce(const void* sendbuff, void* recvbuff, int count,
+                         datatype_t datatype, op_t op,
+                         cudaStream_t stream) const = 0;
+
+  virtual void bcast(void* buff, int count, datatype_t datatype, int root,
+                     cudaStream_t stream) const = 0;
+
+  virtual void reduce(const void* sendbuff, void* recvbuff, int count,
+                      datatype_t datatype, op_t op, int root,
+                      cudaStream_t stream) const = 0;
+
+  virtual void allgather(const void* sendbuff, void* recvbuff, int sendcount,
+                         datatype_t datatype, cudaStream_t stream) const = 0;
+
+  virtual void allgatherv(const void* sendbuf, void* recvbuf,
+                          const int recvcounts[], const int displs[],
+                          datatype_t datatype, cudaStream_t stream) const = 0;
+
+  virtual void reducescatter(const void* sendbuff, void* recvbuff,
+                             int recvcount, datatype_t datatype, op_t op,
+                             cudaStream_t stream) const = 0;
+};
+
+class comms_t : public comms_iface {
```

Comment:
We have an MPI+NCCL  implementation of the comms that allows us to run and test algorithms without the need for bootstrapping UCX in Dask. This allows the handle to hold a comms object without having to be concerned with the implementation. If there’s a better way of doing this in a header-only library, I’m certainly open to alternatives, but I would like it if we kept the implementation of the comms separate from the interface.

Reply:
I am not sure I am following this. Absolutely agree with "we kept the implementation of the comms separate from the interface." What I don't understand is this class has a comms_iface object as a member variable and calls comms_iface functions through this member variable and does nothing else besides this. Anything breaks if we remove "public comms_iface" from the code above?

I understand that std_comms should inherit comms_iface as std_comms implements virtual methods in comms_iface, but I don't get why comms_t here needs to inherit comms_iface.

Reply:
Indeed. I was responding to you on my iphone last night and overlooked this detail. This was an oversight- since there's a conversion constructor from `comms_iface`, this should no longer have to extend `comms_iface`.